### ISSUE TITLE
Use manylinux2014 to provide aarch64 wheels

### ISSUE THREAD

**Issue by hrw** [numpy/numpy#14886]
Manylinux2014 got released with support for wider selection of architectures: aarch64, arm32, ppc64le, s390x. Are there plans to use it for releasing wheels for other architectures than x86-64/i686?

Would cut "pip install numpy" times A LOT of aarch64 platform.

**Comment by mattip** [comment]
xref matthew-brett/multibuild#281

**Comment by hrw** [comment]
https://github.com/hrw/multibuild/tree/hrw-devel has gnu-config change.

for me openblas build fails due to fetching not existing file

**Comment by hrw** [comment]
Multibuild is now aware of manylinux2014. So will next Numpy RC use it? would be great.

**Comment by rgommers** [comment]
I think that would need some discussion. If we release wheels, that's kind of a commitment going forward.

**Comment by hrw** [comment]
Numpy requires several development headers installed + whole toolchain to build. Last time took me a day to figure out all required dependencies. Even then I used in-distro packages so it was still open for several bugs compared to x86-64 wheel version which was built with required versions.

And it required time even on server class hardware. Just to have "pip install numpy" working.

There are several options available to get aarch64 built natively in automated way. All you need to do is integrate them into release/test process and use.

**Comment by rgommers** [comment]
@hrw I know it's helpful, I'm not disagreeing with that. There's a little more to it than you realize though, it's not just "integrate and use".

[PEP 599](https://www.python.org/dev/peps/pep-0599/) specifies the supported architectures as:
```
x86_64
i686
aarch64
armv7l
ppc64
ppc64le
s390x
```
Ones we now have CI for but no wheels are: `aarch64`, `ppc64le` and `s390x`. We need to decide which ones we would like to supply wheels for. And what that means for support policy. Right now the policy for non-`x86_64` platforms can be paraphrased as "we accept patches for these platforms and aim to keep CI green if we have it; we may add `xfail`s to do so". 

Once we decide to publish wheels for a platform, we should probably aim to change that to the same policy as for `x86_64`:
- we publish wheels at release time, before uploading the `sdist`.
- we also make sure we have a compatible OpenBLAS version to build with (through https://github.com/MacPython/openblas-libs) and resolve issues with that as needed.
- we spend effort to resolve issues that are reported (note: this can be hard, because we may not have access to hardware to debug easily).
- we publish nightly wheels for other packages to use in their CI.

Related: for `x86_64` it looks like we should produce only `manylinux2010` wheels until CentOS 6 becomes unsupported (30 Nov 2020), after which we should switch to `manylinux2014`.

**Comment by mattip** [comment]
Access to s390x is more difficult, but ppc64le and aarch64 machines are available via the [gcc compile farm](https://cfarm.tetaneutral.net/machines/list/) to NumPy developers.

The first step in my mind is to extend the MacPython/openblas-libs support to non-x86 architectures, and to use those artifacts in CI runs. 

When we discuss extending our wheel offerings, I would also like to extend the discussion to PyPy as well.

**Comment by rgommers** [comment]
> When we discuss extending our wheel offerings, I would also like to extend the discussion to PyPy as well.

Makes sense, I think the PyPy CI on Azure has been stable and unproblematic for quite a while.

**Comment by mattip** [comment]
> it looks like we should produce only `manylinux2010` wheels

This is still problematic for people with older pip versions. `manylinux2010` is supported with pip 19 and above, which was released less than a year ago.

**Comment by matthew-brett** [comment]
I know this is a judgement call, but switching to manylinux2010 sooner rather than later would be good.  I say that only on the superstitious belief that we run the risk of releasing code with security vulnerabilities by building on a now-unsupported platform.

**Comment by mattip** [comment]
The problem is that pip prefers a sdist tar bundle of a newer version over a wheel of an older version, so suddenly people with an old pip will be trying to build from source. They could do `pip install --binary-only`, but who reads the help.

I think the solution would be to build with manylinux2010, but use a flag to auditwheel to tell it to produce both manylinux1 and manylinux2010 wheels, and upload them both

**Comment by mattip** [comment]
> The first step in my mind is to extend the MacPython/openblas-libs support to non-x86 architectures, and to use those artifacts in CI runs.

The first part is done: there is now support for other architectures in MacPython/openblas-libs. Now we need to 
- create the artifacts for OpenBLAS 0.3.7
- start using them in CI

Then we can begin a PR to add more wheels to MacPython/numpy-wheels

**Comment by mattip** [comment]
The issue of manylinux2014 support in all the supporting packages is solved. The only remaining problem is that manylinux2104 use glibc 2.17, which  triggers blacklisted trigonometric function overrides. These seem to be inaccurate on non-x86 hardware. xref MacPython/numpy-wheels#74

**Comment by odidev** [comment]
[PR#739](https://github.com/pypa/manylinux/pull/739) has fixed the page alignment issue. @mattip, may I know when the next release for numpy is scheduled?

**Comment by mattip** [comment]
@charris is the release manager. I think there may be a 1.19.2 in the works.

**Comment by charris** [comment]
@odidev We can make a 1.19.2 release anytime if there is good reason to do so.

**Comment by charris** [comment]
@odidev Pre-releases for aarch64 were built this afternoon (Sep 3).  You can try them if you wish.
```
pip install -i https://pypi.anaconda.org/scipy-wheels-nightly/simple numpy
```

**Comment by hrw** [comment]
Would be nice to get 1.18 and 1.19 binary wheels updated too.

**Comment by odidev** [comment]
I downloaded the numpy wheel and tried to analyse the readelf output. It looks like the alignment issue is still there in numpy-1.20.0 wheels present at  https://pypi.anaconda.org/scipy-wheels-nightly/simple
ubuntu@1pthunderx1-137:~/FAAS_Project/madhukar/temp/numpy$ readelf -l ./numpy/core/_multiarray_umath.cpython-36m-aarch64-linux-gnu.so

Elf file type is DYN (Shared object file)
Entry point 0x29d40
There are 10 program headers, starting at offset 64

Program Headers:
  Type           Offset             VirtAddr           PhysAddr
                 FileSiz            MemSiz              Flags  Align
  LOAD           0x0000000000000000 0x0000000000000000 0x0000000000000000
                 0x0000000000388620 0x0000000000388620  R E    0x10000
  GNU_STACK      0x0000000000000000 0x0000000000000000 0x0000000000000000
                 0x0000000000000000 0x0000000000000000  RW     0x10
  NOTE           0x0000000000000200 0x0000000000000200 0x0000000000000200
                 0x0000000000000024 0x0000000000000024  R      0x4
  GNU_EH_FRAME   0x00000000003311e0 0x00000000003311e0 0x00000000003311e0
                 0x000000000000bf6c 0x000000000000bf6c  R      0x4
  LOAD           0x000000000038e240 0x000000000039e240 0x000000000039e240
                 0x000000000001f564 0x000000000003fb88  RW     0x10000
  TLS            0x000000000038e240 0x000000000039e240 0x000000000039e240
                 0x0000000000000000 0x0000000000000140  R      0x8
  GNU_RELRO      0x000000000038e240 0x000000000039e240 0x000000000039e240
                 0x0000000000001dc0 0x0000000000001dc0  R      0x1
  LOAD           0x00000000003f9000 0x00000000003de000 0x00000000003de000
                 0x0000000000003a20 0x0000000000003a20  RW     0x1000
  DYNAMIC        0x00000000003fd000 0x00000000003e2000 0x00000000003e2000
                 0x0000000000000220 0x0000000000000220  RW     0x8
  LOAD           0x00000000003fd000 0x00000000003e2000 0x00000000003e2000
                 0x0000000000002eb8 0x0000000000002eb8  RW     0x1000

 Section to Segment mapping:
  Segment Sections...
   00     .dynsym .gnu.version .gnu.version_r .rela.dyn .rela.plt .init .plt .text .fini .rodata .eh_frame_hdr .eh_frame 
   01     
   02     
   03     .eh_frame_hdr 
   04     .init_array .fini_array .data.rel.ro .got .got.plt .data .bss 
   05     .tbss 
   06     .init_array .fini_array .data.rel.ro .got 
   07     .gnu.hash .note.gnu.build-id 
   08     .dynamic 
   09     .dynamic .dynstr 
ubuntu@1pthunderx1-137:~/FAAS_Project/madhukar/temp/numpy$ 


However, I manually build and audited(ran auditwheel repair command) the AArch64 wheel. I can't see alignment issue here. Please have a look at below logs-

Before running auditwheel repair command-
_[root@43e569904f3b dist]# readelf -l ./numpy/core/_multiarray_umath.cpython-36m-aarch64-linux-gnu.so

Elf file type is DYN (Shared object file)
Entry point 0x29e80
There are 8 program headers, starting at offset 64

Program Headers:
  Type           Offset             VirtAddr           PhysAddr
                 FileSiz            MemSiz              Flags  Align
  LOAD           0x0000000000000000 0x0000000000000000 0x0000000000000000
                 0x0000000000301ec8 0x0000000000301ec8  R E    0x10000
  LOAD           0x000000000030e5d8 0x000000000031e5d8 0x000000000031e5d8
                 0x000000000001f1bc 0x000000000003f688  RW     0x10000
  DYNAMIC        0x000000000030fae0 0x000000000031fae0 0x000000000031fae0
                 0x0000000000000220 0x0000000000000220  RW     0x8
  NOTE           0x0000000000000200 0x0000000000000200 0x0000000000000200
                 0x0000000000000024 0x0000000000000024  R      0x4
  TLS            0x000000000030e5d8 0x000000000031e5d8 0x000000000031e5d8
                 0x0000000000000000 0x0000000000000140  R      0x8
  GNU_EH_FRAME   0x00000000002b3690 0x00000000002b3690 0x00000000002b3690
                 0x000000000000c1a4 0x000000000000c1a4  R      0x4
  GNU_STACK      0x0000000000000000 0x0000000000000000 0x0000000000000000
                 0x0000000000000000 0x0000000000000000  RW     0x10
  GNU_RELRO      0x000000000030e5d8 0x000000000031e5d8 0x000000000031e5d8
                 0x0000000000001a28 0x0000000000001a28  R      0x1

 Section to Segment mapping:
  Segment Sections...
   00     .note.gnu.build-id .gnu.hash .dynsym .dynstr .gnu.version .gnu.version_r .rela.dyn .rela.plt .init .plt .text .fini .rodata .eh_frame_hdr .eh_frame 
   01     .init_array .fini_array .data.rel.ro .dynamic .got .got.plt .data .bss 
   02     .dynamic 
   03     .note.gnu.build-id 
   04     .tbss 
   05     .eh_frame_hdr 
   06     
   07     .init_array .fini_array .data.rel.ro .dynamic .got 
[root@43e569904f3b dist]#_ 


Running auditwheel repair command- 
_[root@43e569904f3b dist]# auditwheel repair numpy-1.20.0.dev0+b95ba5d-cp36-cp36m-linux_aarch64.whl 
INFO:auditwheel.main_repair:Repairing numpy-1.20.0.dev0+b95ba5d-cp36-cp36m-linux_aarch64.whl
INFO:auditwheel.wheeltools:Previous filename tags: linux_aarch64
INFO:auditwheel.wheeltools:New filename tags: manylinux2014_aarch64
INFO:auditwheel.wheeltools:Previous WHEEL info tags: cp36-cp36m-linux_aarch64
INFO:auditwheel.wheeltools:New WHEEL info tags: cp36-cp36m-manylinux2014_aarch64
INFO:auditwheel.main_repair:
Fixed-up wheel written to /root/numpy/dist/wheelhouse/numpy-1.20.0.dev0+b95ba5d-cp36-cp36m-manylinux2014_aarch64.whl
[root@43e569904f3b dist]#_ 


After running auditwheel repair command-
_[root@43e569904f3b wheelhouse]# readelf -l ./numpy/core/_multiarray_umath.cpython-36m-aarch64-linux-gnu.so

Elf file type is DYN (Shared object file)
Entry point 0x29e80
There are 8 program headers, starting at offset 64

Program Headers:
  Type           Offset             VirtAddr           PhysAddr
                 FileSiz            MemSiz              Flags  Align
  LOAD           0x0000000000000000 0x0000000000000000 0x0000000000000000
                 0x0000000000301ec8 0x0000000000301ec8  R E    0x10000
  LOAD           0x000000000030e5d8 0x000000000031e5d8 0x000000000031e5d8
                 0x000000000001f1bc 0x000000000003f688  RW     0x10000
  DYNAMIC        0x000000000030fae0 0x000000000031fae0 0x000000000031fae0
                 0x0000000000000220 0x0000000000000220  RW     0x8
  NOTE           0x0000000000000200 0x0000000000000200 0x0000000000000200
                 0x0000000000000024 0x0000000000000024  R      0x4
  TLS            0x000000000030e5d8 0x000000000031e5d8 0x000000000031e5d8
                 0x0000000000000000 0x0000000000000140  R      0x8
  GNU_EH_FRAME   0x00000000002b3690 0x00000000002b3690 0x00000000002b3690
                 0x000000000000c1a4 0x000000000000c1a4  R      0x4
  GNU_STACK      0x0000000000000000 0x0000000000000000 0x0000000000000000
                 0x0000000000000000 0x0000000000000000  RW     0x10
  GNU_RELRO      0x000000000030e5d8 0x000000000031e5d8 0x000000000031e5d8
                 0x0000000000001a28 0x0000000000001a28  R      0x1

 Section to Segment mapping:
  Segment Sections...
   00     .note.gnu.build-id .gnu.hash .dynsym .dynstr .gnu.version .gnu.version_r .rela.dyn .rela.plt .init .plt .text .fini .rodata .eh_frame_hdr .eh_frame 
   01     .init_array .fini_array .data.rel.ro .dynamic .got .got.plt .data .bss 
   02     .dynamic 
   03     .note.gnu.build-id 
   04     .tbss 
   05     .eh_frame_hdr 
   06     
   07     .init_array .fini_array .data.rel.ro .dynamic .got 
[root@43e569904f3b wheelhouse]#_ 


Are you using the latest docker container(manylinux2014_aarch64) to create wheel?

**Comment by mattip** [comment]
[The build](https://travis-ci.org/github/MacPython/numpy-wheels/jobs/723912348) used the docker image with the digest 3e9157a0a00c, which seems to be the one for the [latest arm64 manylinux 2014 docker image](https://quay.io/repository/pypa/manylinux2014_aarch64?tab=tags).

**Comment by odidev** [comment]
The analysis mentioned in [comment](https://github.com/numpy/numpy/issues/14886#issuecomment-687238220) was done on an older numpy wheel (which look to be created before the patchelf fix shipped in [PR #739](https://github.com/pypa/manylinux/pull/739) in manylinux2014_aarch64 containers)

Now again, I picked the latest [numpy wheel](https://pypi.anaconda.org/scipy-wheels-nightly/simple/numpy/1.20.0.dev0+1462f77/numpy-1.20.0.dev0+20200903205810_1462f77-cp36-cp36m-manylinux2014_aarch64.whl) and did the same analysis. Looks like the page alignment issue is gone in the latest release.

@charris, can you please release numpy 1.19.2 to make working wheel available for centos/redhat systems which uses 64kb pages.

**Comment by mattip** [comment]
Closing, we have manylinux2014 wheels for aarch64.
